File: python/string/Anagrams.py
class Solution:
    # @param strs, a list of strings
    # @return a list of strings
    def anagrams(self, strs):
        d = { }
        for i in xrange(len(strs)):
            l = list(strs[i])
            l.sort()
            string = "".join(l)
            if string not in d:
                d[string] = [strs[i]]
            else:
                d[string].append(strs[i])

        result = []
        for key in d:
            if len(d[key]) > 1:
                for j in xrange(len(d[key])):
                    result.append(d[key][j])
        return result


if __name__ == "__main__":
    solution = Solution()

    print solution.anagrams(["", "b"])

File: java/Anagrams.java
public class Solution {
    public List<String> anagrams(String[] strs) {
        Map<String, List<String>> map = new HashMap<String, List<String>>();
        List<String> result = new ArrayList<String>();
        for (int i = 0; i < strs.length; i++) {
            char[] chars = strs[i].toCharArray();
            Arrays.sort(chars);
            String curr = new String(chars);
            if (map.containsKey(curr)) {
                map.get(curr).add(strs[i]);
            } else {
                List<String> valuelist = new ArrayList<String>();
                valuelist.add(strs[i]);
                map.put(curr, valuelist);
            }
        }

        for (Map.Entry<String, List<String>> entry: map.entrySet()) {
            if (entry.getValue().size() > 1) {
                result.addAll(entry.getValue());
            }
        }

        return result;
    }
}
